bridge/bridge.go

```diff
@@ -0,0 +1,224 @@
+	}, nil
+}
+
+func (b *bridge) Start() error {
+	if err := b.transport.Start(); err != nil {
+		return err
+	}
+
+	if err := b.transport.Subscribe(b.addRemoteMessage); err != nil {
+		return err
+	}
+
+	if err := b.tracker.Start(); err != nil {
+		return err
+	}
+
+	eventCh := b.tracker.GetEventChannel()
+	go b.processEvents(eventCh)
+
+	return nil
+}
+
+func (b *bridge) Close() error {
+	close(b.closeCh)
+
+	if err := b.tracker.Stop(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *bridge) SetValidators(validators []types.Address, threshold uint64) {
+	b.resetIsValidatorMap(validators)
+	b.sampool.UpdateValidatorSet(validators, threshold)
+}
+
+func (b *bridge) GetReadyMessages() []sam.ReadyMessage {
+	return b.sampool.GetReadyMessages()
+}
+
+func (b *bridge) Consume(hash types.Hash) {
+	b.sampool.Consume(hash)
+}
+
+func (b *bridge) resetIsValidatorMap(validators []types.Address) {
+	isValidatorMap := make(map[types.Address]bool)
+	for _, address := range validators {
+		isValidatorMap[address] = true
+	}
+
+	b.isValidatorMapLock.Lock()
+	defer b.isValidatorMapLock.Unlock()
+
+	b.isValidatorMap = isValidatorMap
+}
+
+func (b *bridge) isValidator(address types.Address) bool {
+	b.isValidatorMapLock.RLock()
+	defer b.isValidatorMapLock.RUnlock()
+
+	return b.isValidatorMap[address]
+}
+
+func (b *bridge) processEvents(eventCh <-chan []byte) {
+	for {
+		select {
+		case <-b.closeCh:
+			return
+		case data := <-eventCh:
+			if err := b.addLocalMessage(data); err != nil {
+				b.logger.Error("failed process event", "err", err)
+			}
+		}
+	}
+}
+
+func (b *bridge) addLocalMessage(data []byte) error {
+	hash := types.BytesToHash(crypto.Keccak256(data))
+
+	signature, err := b.signer.Sign(hash[:])
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("addLocalMessage hash=%+v,signature=%+v, data=%+v\n", hash, signature, data)
+
+	b.sampool.AddMessage(&sam.Message{
+		Hash: hash,
+		Body: data,
+	})
+
+	b.sampool.AddSignature(&sam.MessageSignature{
+		Hash:      hash,
+		Address:   b.signer.Address(),
+		Signature: signature,
+	})
+
+	signedMessage := &transport.SignedMessage{
+		Hash:      hash,
+		Signature: signature,
+	}
+
+	if err := b.transport.Publish(signedMessage); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *bridge) addRemoteMessage(message *transport.SignedMessage) {
+	sender, err := b.signer.RecoverAddress(message.Hash[:], message.Signature)
+	if err != nil {
+		b.logger.Error("failed to get address from signature", "err", err)
+
+		return
+	}
+
+	fmt.Printf(
+		"addRemoteMessage hash=%+v, signature=%+v, sender=%+v\n",
+		message.Hash,
+		message.Signature,
+		sender,
+	)
+
+	if !b.isValidator(sender) {
+		b.logger.Warn(
+			"ignored gossip message from non-validator",
+			"hash",
+			message.Hash,
+			"from",
+			types.AddressToString(sender),
+		)
+
+		return
+	}
+
+	b.sampool.AddSignature(&sam.MessageSignature{
+		Hash:      message.Hash,
+		Address:   sender,
+		Signature: message.Signature,
+	})
```

Comment:
This doesn't seem good, we shouldn't blindly sign any message we receive from a validator and put it into SAM pool. We should only put it there if we see it first.

Reply:
I don't agree with that. Is it always true that the RootChain event comes first and signature from other validators reach after this?

Reply:
@Kourin1996 You're absolutely right, thanks for the clarification 🎉